vault/identity_store_util.go

```diff
@@ -19,24 +20,57 @@ import (
 	"github.com/hashicorp/vault/logical"
 )
 
+var (
+	errDuplicateIdentityName = errors.New("duplicate identity name")
+)
+
 func (c *Core) loadIdentityStoreArtifacts(ctx context.Context) error {
-	var err error
 	if c.identityStore == nil {
 		c.logger.Warn("identity store is not setup, skipping loading")
 		return nil
 	}
 
-	err = c.identityStore.loadEntities(ctx)
-	if err != nil {
+	loadFunc := func(context.Context) error {
+		err := c.identityStore.loadEntities(ctx)
+		if err != nil {
+			return err
+		}
+		return c.identityStore.loadGroups(ctx)
+	}
+
+	// Load everything when memdb is set to operate on lower cased names
+	err := loadFunc(ctx)
+	switch {
+	case err == nil:
+		// If it succeeds, all is well
+		return nil
+	case err != nil && !errwrap.ContainsType(err, errDuplicateIdentityName):
```

Comment:
I don't think ContainsType does what you think it does here, which makes me wonder if this works properly. ContainsType is supposed to be used to validate a typed error, which means this will likely match against any error generated via `errors.New`. You probably want `.Contains()`

Reply:
Wow. This is a nice observation. Fixed. But the code was working properly because the code expecting a `true` and ContainsType (as you said) was returning `true`. There wasn't a different error there to catch. I tested it by overwriting the `err` returned from `loadFunc` and verified that its getting caught properly now.